Astro Quickstart (#2021)
astronomer/docs: astro/create-dag.md

    │   └── test_dag_integrity.py # Test that checks for basic errors in your DAGs
    ├── airflow_settings.yaml # For your Airflow connections, variables and pools (local only)
    ├── packages.txt # For OS-level packages
    └── requirements.txt # For Python packages
    ```

    This set of files builds a Docker image that you can both run on your local machine and deploy to Astro.

## Step 2: Build your project locally

To confirm that you successfully initialized an Astro project, run the following command from your project directory:

```sh
astro dev start
```

This command builds your project and spins up 4 Docker containers on your machine, each for a different Airflow component:

- **Postgres:** Airflow's metadata database
- **Webserver:** The Airflow component responsible for rendering the Airflow UI
- **Scheduler:** The Airflow component responsible for monitoring and triggering tasks
- **Triggerer:** The Airflow component responsible for running Triggers and signaling tasks to resume when their conditions have been met. The triggerer is used exclusively for tasks that are run with [deferrable operators](https://docs.astronomer.io/learn/deferrable-operators)

> If your project builds successfully, the Airflow UI automatically opens in your default webserver at `https://localhost:8080/`.

Reviewer: ```suggestion
It might take a few minutes for your project to build. When your project builds successfully, the Airflow UI automatically opens in your default webserver at `https://localhost:8080/`.
```

> When you create an Astro project, the Astro CLI uses port `8080` for the Airflow webserver and port `5432` for the Airflow metadata database by default. If these ports are already in use on your local computer, an error message might appear. To resolve this error message, see [Test and troubleshoot locally](test-and-troubleshoot-locally.md#ports-are-not-available).
> 
> ## Step 3: Locally access the Airflow UI
> 
> After your project is running, you can access the Airflow UI by going to `http://localhost:8080/` and logging in with `admin` for both your username and password.
> 
> :::info
> 
> It might take a few minutes for the Airflow UI to be available. As you wait for the webserver container to start up, you may need to refresh your browser.
> 
> :::
> 
> After logging in, you should see the DAGs from your `dags` directory in the Airflow UI.
> 
> ![Example DAG in the Airflow UI](/img/docs/sample-dag.png)

Reviewer: ```suggestion
On the Airflow UI home page, you should see the DAGs from your `dags` directory in the Airflow UI.

![Example DAG in the Airflow UI](/img/docs/sample-dag.png)

When you create an Astro project, the Astro CLI uses port `8080` for the Airflow webserver and port `5432` for the Airflow metadata database by default. If these ports are already in use on your local computer, an error message might appear. To resolve this error message, see [Test and troubleshoot locally](test-and-troubleshoot-locally.md#ports-are-not-available).
```

## Step 4: Authenticate to Astro

> Now that you've created your Airflow project and confirmed that your project runs Airflow locally, you can log in to Astro from the CLI.

Reviewer: Can we reword this so that it doesn't sound like having a project is a prerequisite for logging into the Astro CLI? It's not necessarily even  a prerequisite for deploying the project, as running the default project locally is more of a test of your machine than the Astro project itself.

Author: I think actually this clarification makes more sense in the *Build your project locally* step - explaining why to build locally, so I added it there. Also, as I am a very typical user who jumps around in quickstarts all the time, having the "this is technically an optional step" information in the actual optional step is really helpful if something isn't working as I expected, and it seems like there can be some troubleshooting involved if the ports are configured a certain way.

Run the following command to authenticate to Astro:

```sh
astro login
```

After running this command, you are prompted to open your web browser and log in to the Cloud UI. Then, you are automatically authenticated to the CLI.

> ## Step 5: Access the Astro Cloud UI

Reviewer: ```suggestion
## Step 5: Access the Cloud UI
```

> 1. Go to `https://cloud.astronomer.io`, and select one of the following options to access the Cloud UI:
> 
>     - To authenticate with single sign-on (SSO), enter your email and click **Continue**. If your Organization has SSO enabled, you'll be redirected to your identity provider authentication screen.
>     - To authenticate with your GitHub account, click **Continue with GitHub**, enter your username or email address, enter your password, and then click **Sign in**. If your Organization selects this log in option, you’ll receive an email invitation from your Organization Owner. You can't access the Organization without an invitation.
>     - To authenticate with your Google account, click **Continue with Google**, choose an account, enter your username and password, and then click **Sign In**. If your Organization selects this log in option, you’ll receive an email invitation from your Organization Owner. You can't access the Organization without an invitation.

Reviewer: ```suggestion
1. Go to `https://cloud.astronomer.io`, and select one of the available options to access the Cloud UI.
```
I think these descriptions are out of date 😓 you don't necessarily need an invite anymore, gotta remember to update the login doc. But also, Organization can disable these methods so the options don't apply to all users. Maybe it's better to just be general here?

## Step 6: Create a deployment

1. In the Cloud UI, select a **Workspace**.

2. On the **Deployments** page, click **Deployment**.

3. Complete the following fields:

    - **Name**: Enter a name for your Deployment.
    - **Cluster**: Select the Astro cluster in which you want to create this Deployment.
  

>   You can leave the other fields at their default values.

Reviewer: ```suggestion
    You can leave the other fields at their default values.
```

Reviewer: This would be a good time to assure users that we won't be running up a crazy tab for the resources they'll be using

4. Click **Create Deployment**.

     A confirmation message appears indicating that the Deployment is in progress. Select the **Deployments** link to go to the **Deployments** page. The Deployment status is **Creating** until all underlying components in your Astro cluster are healthy, including the Airflow webserver and scheduler. During this time, the Airflow UI is unavailable and you can't deploy code or modify Deployment settings. When the Deployment is ready, the status changes to **Healthy**.
    
    For more information about possible Deployment health statuses, see [Deployment health](deployment-metrics.md#deployment-health).

> ## Step 7: Deploy your Astro Project to your Astro Deployment

Reviewer: ```suggestion
## Step 7: Deploy your Astro project to your Astro Deployment
```

To provide a basic demonstration of an ETL pipeline, this DAG creates an example JSON string, calculates a value based on the string, and prints the results of the calculation to the Airflow logs.

Before you can run any DAG in Astro, you have to unpause it in your remote Airflow instance and then manually trigger a DAG run. 

1. Open Airflow in the Astro UI. To unpause `example-dag-basic`, click the slider button next to its name, and the DAG starts to run on the schedule defined in its code.

    ![Pause DAG slider in the Airflow UI](/img/docs/tutorial-unpause-dag.png)

2. While all DAGs can run on a schedule defined in their code, you can manually trigger a DAG run at any time from the Airflow UI. Manually trigger `example-dag-basic` by clicking the play button under the **Actions** column. During development, running DAGs on demand can help you identify and resolve issues.

After you press **Play**, the **Runs** and **Recent Tasks** sections for the DAG start to populate with data.

![DAG running in the Airflow UI](/img/docs/tutorial-run-dag.png)

> These circles represent different [states](https://airflow.apache.org/docs/apache-airflow/stable/core-concepts/tasks.html#task-instances) that your DAG and task runs can be in. However, these are only high-level summaries of your runs that won't make much sense until you learn more about how Airflow works. 

Reviewer: ```suggestion
```
I think we could use a "Next steps" section here, or possibly another step that tells users to check the Cloud UI for additional metrics about their Airflow task runs?

id: 'create-first-DAG'
---

<head>
  <meta name="description" content="Learn how to deploy and run a DAG in an Astro project with the Astro command-line interface (CLI). You need an Astro project to run Airflow pipelines on Astro." />
  <meta name="og:description" content="Learn how to deploy and run a DAG in an Astro project with the Astro command-line interface (CLI). You need an Astro project to run Airflow pipelines on Astro." />
</head>

import {siteVariables} from '@site/src/versions';

This quickstart gives you a demonstration for how to run your first DAG on Astro by building a sample Astro project, deploying it to the Astro Cloud, and then triggering a DAG run.

To run an Airflow DAG on Astro, you need to complete the following:

- Create an _Astro project_. An Astro project contains the set of files necessary to run Airflow, including dedicated folders for your DAG files, plugins, and dependencies. After you've tested these files locally, the Astro project structure makes it easy to deploy your pipelines to Astro.
- Authenticate and log in to Astro. 
- Create a Deployment to host your Airflow environment.
- Deploy your project to Astro.

> - Trigger your DAG on Astro. 

Author: I think in this case, it makes more sense to have the full definitions in the context of the step. I was on the fence when I was looking at the **Create a project** step, and it really helps orient and teach to have it all at the same time. Seeing a short bullet list of "things we're doing" at the beginning is kind of a summary introduction. I don't know that we're teaching at this point in the quickstart, while we are definitely teaching later on.

> To run Airflow pipelines on Astro, you need to complete the following:
> 
> - Create an Astro project. An Astro project contains the set of files necessary to run Airflow, including dedicated folders for your DAG files, plugins, and dependencies. After you've tested these files locally, the Astro project structure makes it easy to deploy your pipelines to Astro.
> - Authenticate and log in to Astro. 
> - Create a Deployment to host your Airflow environment.
> - Deploy your project to Astro.
> - Trigger your DAG on Astro. 
> 
> This quickstart gives you a demonstration for how to run your first DAG on Astro by building a sample Astro project, deploying it to the Astro Cloud, and then triggering a DAG run, which is an instance of a DAG running at a specifically scheduled time. 

Reviewer: ```suggestion
This quickstart gives you a demonstration for how to run your first DAG on Astro by building a sample Astro project, deploying it to the Astro Cloud, and then triggering a DAG run.

To run an Airflow DAG on Astro, you need to complete the following:

- Create an _Astro project_. An Astro project contains the set of files necessary to run Airflow, including dedicated folders for your DAG files, plugins, and dependencies. After you've tested these files locally, the Astro project structure makes it easy to deploy your pipelines to Astro.
- Authenticate and log in to Astro. 
- Create a Deployment to host your Airflow environment.
- Deploy your project to Astro.
- Trigger your DAG on Astro. 
```

## Prerequisites

- The [Astro CLI](cli/install-cli.md)
- [Docker](https://www.docker.com/products/docker-desktop)
- An Astronomer account. To create your own account, see [Start your Astro trial](astro/trial.md).
- A [cluster](create-cluster.md) and a [Workspace](manage-workspaces.md).

## Step 1: Create an Astro project

> 1. Create a new directory for your Astro project:

Author: This is a good question - how beginner are we considering beginner? beginner to coding vs beginner to Astro/Airflow (with some development experience)

3. Run the following Astro CLI command to initialize an Astro project in the directory:

    ```sh
    astro dev init
    ```

    This command generates the following files in the directory:

    ```
    .
    ├── .env # Local environment variables
    ├── dags # Where your DAGs go
    │   ├── example-dag-basic.py # Example DAG that showcases a simple ETL data pipeline
    │   └── example-dag-advanced.py # Example DAG that showcases more advanced Airflow features, such as the TaskFlow API
    ├── Dockerfile # For the Astro Runtime Docker image, environment variables, and overrides
    ├── include # For any other files you'd like to include
    ├── plugins # For any custom or community Airflow plugins
    │   └── example-plugin.py
    ├── tests # For any DAG unit test files to be run with pytest
    │   └── test_dag_integrity.py # Test that checks for basic errors in your DAGs
    ├── airflow_settings.yaml # For your Airflow connections, variables and pools (local only)
    ├── packages.txt # For OS-level packages
    └── requirements.txt # For Python packages
    ```

    This set of files builds a Docker image that you can both run on your local machine and deploy to Astro.

> ## Step 2: (Optional) Build your project locally

Reviewer 2: I'm going back and forth on making this optional... let's maybe remove that part? If we think people should do it, we include it in the doc -- if we don't, let's remove

(I say keep/make people do it)

> ## Step 2: Build your project locally

Author: ```suggestion
## Step 2: (Optional) Build your project locally
```

> Building your project locally allows you to test your DAGs locally and confirm that your example Astro project builds correctly before you run it remotely in Astro.

Reviewer: ```suggestion
Building your project locally allows you to test your DAGs locally and confirm that your example Astro project builds correctly before you run it remotely in Astro. While this step is not required for deploying and running your code on Astro, Astronomer recommends always using the Astro CLI to test locally before deploying.
```
We could also add "Optional" to the step title, but I'm OK either way on that

To start running your project in a local Airflow environment, run the following command from your project directory:

```sh
astro dev start
```

This command builds your project and spins up 4 Docker containers on your machine, each for a different Airflow component:

- **Postgres:** Airflow's metadata database
- **Webserver:** The Airflow component responsible for rendering the Airflow UI
- **Scheduler:** The Airflow component responsible for monitoring and triggering tasks
- **Triggerer:** The Airflow component responsible for running Triggers and signaling tasks to resume when their conditions have been met. The triggerer is used exclusively for tasks that are run with [deferrable operators](https://docs.astronomer.io/learn/deferrable-operators)

> After your project builds successfully, the Airflow UI automatically opens in your default webserver at `https://localhost:8080/`.

Reviewer: ```suggestion
After your project builds successfully, the Airflow UI automatically opens in your default web browser at `https://localhost:8080/`.
```

    
    For more information about possible Deployment health statuses, see [Deployment health](deployment-metrics.md#deployment-health).

## Step 5: Deploy your Astro project to your Astro Deployment

To deploy your Astro project, run:

```sh
astro deploy
```

This command returns a list of Deployments available in your Workspace and prompts you to pick one.

After you select a Deployment, the CLI parses your DAGs to ensure that they don't contain basic syntax and import errors. This test is equivalent to the one that runs during `astro dev parse` in a local Airflow environment. If any of your DAGs fail this parse, the deploy to Astro also fails.

If your code passes the parse, the Astro CLI builds all files in your Astro project directory into a new Docker image and then pushes the image to your Deployment on Astro. If the DAG-only deploy feature is enabled for your Deployment, the `/dags` directory is excluded from this Docker image and pushed separately. To force a deploy even if your project has DAG errors, you can run `astro deploy --force`.

## Step 6: Trigger your DAG on Astro

> Before you can run any DAG in Astro, unpause it in your remote Airflow instance and then manually trigger a DAG run. 

Reviewer 2: ```suggestion
When you deploy your Astro project, all example DAGs will initially be paused and will not start running automatically. To run one of these example DAGs according to its schedule, you have to unpause it from the Airflow UI.
```

Tryna give the user more logical end-to-end context here

> 1. Select the **Open Airflow** button in the Astro UI to open the Airflow dashboard. 

Reviewer 2: ```suggestion
1. In the Deployment page of the Cloud UI, click the **Open Airflow** button.
```

- Always Cloud UI! In style guide
- For a button, I think we say `click` and we keep `select` when there's a dropdown or a. menu of options

> 2. Click the slider button next to `example-dag-basic` to unpause it, and the DAG starts to run on the schedule defined in its code.

Reviewer 2: ```suggestion
2. In the main DAGs view of the Airflow UI, click the slider button next to `example-dag-basic` to unpause it. If you hover over the DAG, it should say `DAG is Active`. When you do this, the DAG starts to run on the schedule that is defined in its code.
```

    ![Pause DAG slider in the Airflow UI](/img/docs/tutorial-unpause-dag.png)

> 3. While all DAGs can run on a schedule defined in their code, you can manually trigger a DAG run at any time from the Airflow UI. Manually trigger `example-dag-basic` by clicking the play button under the **Actions** column. During development, running DAGs on demand can help you identify and resolve issues.

Reviewer 2: ```suggestion
3. Manually trigger a DAG run of `example-dag-basic` by clicking the play button under the **Actions** column. This schedules all task instances in the DAG in addition to its schedule. When you develop DAGs on Astro, triggering a DAG run instead of waiting for the DAG schedule can help you quickly identify and resolve issues.
```

Not perfect, but in gen I think we can do better wording here

Author: Do we need to bring up task instances at this point in the intro/quickstart? We haven't introduced the concept before this step - unless they're already familiar with DAGs.

    After you press **Play**, the **Runs** and **Recent Tasks** sections for the DAG start to populate with data.

    ![DAG running in the Airflow UI](/img/docs/tutorial-run-dag.png)

>     These circles represent different [states](https://airflow.apache.org/docs/apache-airflow/stable/core-concepts/tasks.html#task-instances) that your DAG and task runs can be in. However, these are only high-level summaries of your runs.

Reviewer 2: Consider wording here -- what does "these are only high-level summaries of your runs"? mean exactly

Author: Removed this sentence - I don't think it's necessary detail for a quickstart.

> 2. While all DAGs can run on a schedule defined in their code, you can manually trigger a DAG run at any time from the Airflow UI. Manually trigger `example-dag-basic` by clicking the play button under the **Actions** column. During development, running DAGs on demand can help you identify and resolve issues.
> 
> After you press **Play**, the **Runs** and **Recent Tasks** sections for the DAG start to populate with data.
> 
> ![DAG running in the Airflow UI](/img/docs/tutorial-run-dag.png)
> 
> These circles represent different [states](https://airflow.apache.org/docs/apache-airflow/stable/core-concepts/tasks.html#task-instances) that your DAG and task runs can be in. However, these are only high-level summaries of your runs.
> 
> 3. Click on the name of the DAG, **example-dag-basic** to open a detailed view in Airflow about the DAG, then select **Audit Log** to see a record of the actions performed and its run history. To see if your DAG ran successfully, the final event in the Audit Log should be a `success` for a `load` task.
> 
> 4. After you finish testing your DAGs in your Deployment, pause your DAG by opening Airflow and clickincing the slider button next to `example-dag-basic`. This prevents your example DAG from running automatically, so it won't consume resources unexpectedly.

Reviewer: ```suggestion
3. While all DAGs can run on a schedule defined in their code, you can manually trigger a DAG run at any time from the Airflow UI. Manually trigger `example-dag-basic` by clicking the play button under the **Actions** column. During development, running DAGs on demand can help you identify and resolve issues.

    After you press **Play**, the **Runs** and **Recent Tasks** sections for the DAG start to populate with data.

    ![DAG running in the Airflow UI](/img/docs/tutorial-run-dag.png)

    These circles represent different [states](https://airflow.apache.org/docs/apache-airflow/stable/core-concepts/tasks.html#task-instances) that your DAG and task runs can be in. However, these are only high-level summaries of your runs.

4. Click on the name of the DAG, **example-dag-basic**, to open the **Grid** view for the DAG. To see if your DAG ran successfully, the most recent entry in the grid should have green squares for all of your tasks.

5. After you finish testing your DAGs in your Deployment, pause your DAG by opening Airflow and clicking the slider button next to `example-dag-basic`. This prevents your example DAG from running automatically, so it won't consume resources unexpectedly.
```

> 5. After you finish testing your DAGs in your Deployment, pause your DAG by opening Airflow and clicking the slider button next to `example-dag-basic`. This prevents your example DAG from running automatically, so it won't consume resources unexpectedly.

Reviewer 2: ```suggestion
5. Pause your DAG by clicking the slider button next to `example-dag-basic`. This prevents your example DAG from running automatically and consuming your Deployment resources.
```

- You weren't really testing this DAG
- "by opening Airflow" -- Airflow was already open?

> ## Step 7: View your DAG status in Astro

Reviewer 2: ```suggestion
## Step 7: View your DAG status in the Cloud UI
```

i might specify cloud UI here since you're hopping between UIs and it could be confusing

> 1. Open your Astro Workspace. Since you ran your example DAG, your Workspace has summary information about your DAG runs and the health of your Deployment available.

Reviewer: ```suggestion
Open your Astro Deployment. Because you ran your example DAG, your Deployment information page now has summary information about your DAG runs and the health of your Deployment.
```

> Open your Astro Deployment. Because you ran your example DAG, your Deployment information page now has summary information about your DAG runs and the health of your Deployment.

Reviewer 2: - Would be good to add 1 statement about what we expect you to do in Cloud UI vs. Airflow UI
- Is this the analytics page?

Author: * Took a stab at a summary
* This is a screenshot of the analytics page - this info is completely blank when you set it up as a quickstart and it's an immediate confirmation of "hey it worked! there's stuff there!" that I noticed so I wanted to give it a little highlight. If there's a better choice, happy to change or tweak this status.

![Summary information about your DAG runs in Astro UI](/img/docs/first-DAG-data.png)

## Step 8: (Optional) Delete your Deployment

> You might want to delete your Deployment after you finish triggering your DAG test runs and exploring Astro. 

Reviewer: ```suggestion
To limit resource usage, you might want to delete your Deployment after you finish triggering your DAG test runs.
```

> 1. Open the **Deployments** page in Astro and select your Deployment. 

Reviewer: ```suggestion
1. In the Cloud UI, open your Workspace, then open your Deployment.
```
Weird one in our style guide, but we always try to refer to the UI as the "Cloud UI" instead of "Astro", since Astro embodies a lot more than just the UI

> 2. Select the ellipses to see the more options, and select **Delete**.

Reviewer: ```suggestion
2. Select the ellipses to see more options, then click **Delete**.
```

3. When prompted, confirm the deletion by typing **DELETE**.

## Next Steps

Now that you've created and run your first DAG on Astro, you can learn more about working with Astro, Airflow, and DAGs.

- [Develop a project](/astro/develop-project)
- [Learn how to write your own DAGs](https://docs.astronomer.io/learn/category/dags)
- [Deploy code to Astro](/astro/deploy-code)

> - [Set up CI/CD](/astro/set-up-ci-cd)

Reviewer 2: Perfect links!
